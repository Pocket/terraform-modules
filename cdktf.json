{
  "language": "typescript",
  "app": "npm run --silent compile && node dist/src/example.js",
  "terraformProviders": [
    "pagerduty/pagerduty@~> 1.8.0",
    "archive@~> 1.3.0",
    "local@~> 2.1.0"
  ]
}
