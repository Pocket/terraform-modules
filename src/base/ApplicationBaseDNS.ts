import { Resource } from 'cdktf';
import {
  DataAwsRoute53Zone,
  Route53Record,
  Route53Zone,
} from '../../.gen/providers/aws';
import { Construct } from 'constructs';
import { getRootDomain } from '../utilities';

export interface RootDNSProps {
  domain: string;
  tags?: { [key: string]: string };
}

export class ApplicationBaseDNS extends Resource {
  public readonly zoneId: string;

  constructor(scope: Construct, name: string, config: RootDNSProps) {
    super(scope, name);

    const route53MainZone = new DataAwsRoute53Zone(
      scope,
      `${name}_main_hosted_zone`,
      {
        name: getRootDomain(config.domain),
        tags: config.tags,
      }
    );

    const route53SubZone = new Route53Zone(this, `${name}_subhosted_zone`, {
      name: config.domain,
      tags: config.tags,
    });

    this.zoneId = route53SubZone.zoneId;

    new Route53Record(this, `${name}_subhosted_zone_ns`, {
      name: config.domain,
      type: 'NS',
      ttl: 86400,
      zoneId: route53MainZone.zoneId,
      records: route53SubZone.nameServers,
    });
  }
}
