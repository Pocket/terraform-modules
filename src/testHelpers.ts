import { Resource, TerraformStack } from 'cdktf';
import { Construct } from 'constructs';

// used to test individual static functions within component classes
// generates a basic/empty construct Resource to provide context
export class TestResource extends Resource {
  constructor(scope: TerraformStack | Construct, name: string) {
    super(scope, name);
  }
}
