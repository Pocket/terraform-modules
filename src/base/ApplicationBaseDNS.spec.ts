import { Testing, TerraformStack } from 'cdktf';
import { TestResource } from '../testHelpers';
import { ApplicationBaseDNS } from './ApplicationBaseDNS';

describe('ApplicationBaseDNS', () => {
  let app;
  let stack;

  const tags = {
    name: 'thedude',
    hobby: 'bowling',
  };

  beforeEach(() => {
    app = Testing.app();
    stack = new TerraformStack(app, 'test');
  });

  describe('retrieveAwsRoute53Zone', () => {
    it('retrieves a route 53 zone', () => {
      const construct = new TestResource(stack, 'test-resource');

      ApplicationBaseDNS.retrieveAwsRoute53Zone(
        construct,
        'some-zone',
        'dev.gobowling.info'
      );

      expect(Testing.synth(stack)).toMatchSnapshot();
    });
  });

  describe('generateRoute53Zone', () => {
    it('renders a route 53 zone', () => {
      const construct = new TestResource(stack, 'test-resource');

      ApplicationBaseDNS.generateRoute53Zone(construct, 'dev.gobowling.info');

      expect(Testing.synth(stack)).toMatchSnapshot();
    });

    it('renders a route 53 zone with tags', () => {
      const construct = new TestResource(stack, 'test-resource');

      ApplicationBaseDNS.generateRoute53Zone(
        construct,
        'dev.gobowling.info',
        tags
      );

      expect(Testing.synth(stack)).toMatchSnapshot();
    });
  });

  describe('generateRoute53Record', () => {
    it('renders a route 53 record', () => {
      const construct = new TestResource(stack, 'test-resource');

      ApplicationBaseDNS.generateRoute53Record(
        construct,
        'dev.gobowling.info',
        'some-zone-id',
        ['some', 'records']
      );

      expect(Testing.synth(stack)).toMatchSnapshot();
    });
  });

  describe('constructor', () => {
    it('renders base DNS without tags', () => {
      new ApplicationBaseDNS(stack, 'testDNS', {
        domain: 'dev.gobowling.info',
      });

      expect(Testing.synth(stack)).toMatchSnapshot();
    });

    it('renders base DNS with tags', () => {
      new ApplicationBaseDNS(stack, 'testDNS', {
        domain: 'dev.gobowling.info',
        tags: tags,
      });

      expect(Testing.synth(stack)).toMatchSnapshot();
    });
  });
});
