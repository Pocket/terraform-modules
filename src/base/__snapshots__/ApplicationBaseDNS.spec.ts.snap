// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApplicationBaseDNS constructor renders base DNS with tags 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"test\\"
    }
  },
  \\"resource\\": {
    \\"aws_route53_zone\\": {
      \\"testDNS_subhosted_zone_0D42EC63\\": {
        \\"name\\": \\"dev.gobowling.info\\",
        \\"tags\\": {
          \\"name\\": \\"thedude\\",
          \\"hobby\\": \\"bowling\\"
        },
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testDNS/subhosted_zone\\",
            \\"uniqueId\\": \\"testDNS_subhosted_zone_0D42EC63\\"
          }
        }
      }
    },
    \\"aws_route53_record\\": {
      \\"testDNS_subhosted_zone_ns_5F02B4AB\\": {
        \\"name\\": \\"dev.gobowling.info\\",
        \\"records\\": \\"\${aws_route53_zone.testDNS_subhosted_zone_0D42EC63.name_servers}\\",
        \\"ttl\\": 86400,
        \\"type\\": \\"NS\\",
        \\"zone_id\\": \\"\${data.aws_route53_zone.testDNS_main_hosted_zone.zone_id}\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testDNS/subhosted_zone_ns\\",
            \\"uniqueId\\": \\"testDNS_subhosted_zone_ns_5F02B4AB\\"
          }
        }
      }
    }
  },
  \\"data\\": {
    \\"aws_route53_zone\\": {
      \\"testDNS_main_hosted_zone\\": {
        \\"name\\": \\"gobowling.info\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testDNS_main_hosted_zone\\",
            \\"uniqueId\\": \\"testDNS_main_hosted_zone\\"
          }
        }
      }
    }
  }
}"
`;

exports[`ApplicationBaseDNS constructor renders base DNS without tags 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"test\\"
    }
  },
  \\"resource\\": {
    \\"aws_route53_zone\\": {
      \\"testDNS_subhosted_zone_0D42EC63\\": {
        \\"name\\": \\"dev.gobowling.info\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testDNS/subhosted_zone\\",
            \\"uniqueId\\": \\"testDNS_subhosted_zone_0D42EC63\\"
          }
        }
      }
    },
    \\"aws_route53_record\\": {
      \\"testDNS_subhosted_zone_ns_5F02B4AB\\": {
        \\"name\\": \\"dev.gobowling.info\\",
        \\"records\\": \\"\${aws_route53_zone.testDNS_subhosted_zone_0D42EC63.name_servers}\\",
        \\"ttl\\": 86400,
        \\"type\\": \\"NS\\",
        \\"zone_id\\": \\"\${data.aws_route53_zone.testDNS_main_hosted_zone.zone_id}\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testDNS/subhosted_zone_ns\\",
            \\"uniqueId\\": \\"testDNS_subhosted_zone_ns_5F02B4AB\\"
          }
        }
      }
    }
  },
  \\"data\\": {
    \\"aws_route53_zone\\": {
      \\"testDNS_main_hosted_zone\\": {
        \\"name\\": \\"gobowling.info\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/testDNS_main_hosted_zone\\",
            \\"uniqueId\\": \\"testDNS_main_hosted_zone\\"
          }
        }
      }
    }
  }
}"
`;

exports[`ApplicationBaseDNS generateRoute53Record renders a route 53 record 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"test\\"
    }
  },
  \\"resource\\": {
    \\"aws_route53_record\\": {
      \\"test-resource_subhosted_zone_ns_47594402\\": {
        \\"name\\": \\"dev.gobowling.info\\",
        \\"records\\": [
          \\"some\\",
          \\"records\\"
        ],
        \\"ttl\\": 86400,
        \\"type\\": \\"NS\\",
        \\"zone_id\\": \\"some-zone-id\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/test-resource/subhosted_zone_ns\\",
            \\"uniqueId\\": \\"test-resource_subhosted_zone_ns_47594402\\"
          }
        }
      }
    }
  }
}"
`;

exports[`ApplicationBaseDNS generateRoute53Zone renders a route 53 zone 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"test\\"
    }
  },
  \\"resource\\": {
    \\"aws_route53_zone\\": {
      \\"test-resource_subhosted_zone_262A1A23\\": {
        \\"name\\": \\"dev.gobowling.info\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/test-resource/subhosted_zone\\",
            \\"uniqueId\\": \\"test-resource_subhosted_zone_262A1A23\\"
          }
        }
      }
    }
  }
}"
`;

exports[`ApplicationBaseDNS generateRoute53Zone renders a route 53 zone with tags 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"test\\"
    }
  },
  \\"resource\\": {
    \\"aws_route53_zone\\": {
      \\"test-resource_subhosted_zone_262A1A23\\": {
        \\"name\\": \\"dev.gobowling.info\\",
        \\"tags\\": {
          \\"name\\": \\"thedude\\",
          \\"hobby\\": \\"bowling\\"
        },
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/test-resource/subhosted_zone\\",
            \\"uniqueId\\": \\"test-resource_subhosted_zone_262A1A23\\"
          }
        }
      }
    }
  }
}"
`;

exports[`ApplicationBaseDNS retrieveAwsRoute53Zone retrieves a route 53 zone 1`] = `
"{
  \\"//\\": {
    \\"metadata\\": {
      \\"version\\": \\"stubbed\\",
      \\"stackName\\": \\"test\\"
    }
  },
  \\"data\\": {
    \\"aws_route53_zone\\": {
      \\"test-resource_some-zone_main_hosted_zone_B55D697C\\": {
        \\"name\\": \\"gobowling.info\\",
        \\"//\\": {
          \\"metadata\\": {
            \\"path\\": \\"test/test-resource/some-zone_main_hosted_zone\\",
            \\"uniqueId\\": \\"test-resource_some-zone_main_hosted_zone_B55D697C\\"
          }
        }
      }
    }
  }
}"
`;
